### PR TITLE
Add CUDA atomic.{max, min, compare_and_swap}

### PR DESCRIPTION
**Pull request by sklam**
* Extends min, max for  float64, float32, int32, int64, uint32, uint64
* Add compare_and_swap for int32

### CODE REVIEW

File: numba/cuda/cudadecl.py [hunk 1]
```diff
@@ -119,17 +119,16 @@ def generic(self, args, kws):
...
     def generic(self, args, kws):
         assert not kws
         ary, idx, val = args
-
-        # Implementation presently supports float64 only,
+        # Implementation presently supports:
+        # float64, float32, int32, int64, uint32, uint64 only,
         # so fail typing otherwise
-        if ary.dtype != types.float64:
+        supported_types = (types.float64, types.float32, types.int32,
```
Comment (pitrou) [inline numba/cuda/cudadecl.py]: Just curious, CUDA doesn't support this on smaller types (e.g. int16)?

File: numba/cuda/cudadrv/nvvm.py [hunk 2]
```diff
@@ -384,34 +384,65 @@ def get(self):
...
+define internal {T} @___numba_atomic_{T}_min({T}* %ptr, {T} %val) alwaysinline{{
+entry:
+    %ptrval = load volatile {T}, {T}* %ptr
+    ; Check if val is a NaN and return *ptr early if so
+    %valnan = fcmp uno {T} %val, %val
+    br i1 %valnan, label %done, label %gt_check
+
+gt_check:
+    %dold = phi {T} [ %ptrval, %entry ], [ %dcas, %attempt ]
+    ; Continue attempts if dold < val or dold is NaN (using ult semantics)
```
Comment (pitrou) [inline numba/cuda/cudadrv/nvvm.py]: `dold > val` and `ugt`

File: numba/cuda/cudadrv/nvvm.py [hunk 3]
```diff
@@ -384,34 +384,65 @@ def get(self):
...
+    ; Attempt to swap in the smaller value
+    %iold = bitcast {T} %dold to {Ti}
+    %iptr = bitcast {T}* %ptr to {Ti}*
+    %ival = bitcast {T} %val to {Ti}
+    %cas = cmpxchg volatile {Ti}* %iptr, {Ti} %iold, {Ti} %ival monotonic
+    %dcas = bitcast {Ti} %cas to {T}
+    br label %gt_check
+
+done:
+    ; Return max
```
Comment (pitrou) [inline numba/cuda/cudadrv/nvvm.py]: "Return min"

File: numba/cuda/cudadecl.py (hunk 1 above)
Comment (seibert) [inline numba/cuda/cudadecl.py]: Correct.  Each of these atomic functions have to be supported in hardware (the speculation is that support even needs to be present in the on-chip memory controller), so adding support for types happens very slowly.  Full support for 64-bit only happened in the last few architecture generations.

File: numba/cuda/tests/cudapy/test_atomics.py [hunk 4]
```diff
@@ -261,15 +289,114 @@ def test_atomic_add_double_global_3(self):
 
         np.testing.assert_equal(ary, orig + 1)
 
+    def test_atomic_max_int32(self):
```
Comment (pitrou) [inline numba/cuda/tests/cudapy/test_atomics.py]: There's a lot of repetition amongst the tests for various dtypes, perhaps the repetition can be factored out?

File: numba/cuda/stubs.py [hunk 5]
```diff
@@ -308,5 +308,25 @@ class max(Stub):
         differs from Python and Numpy behaviour, where max(a, b) is always
         a when either a or b is a NaN.
 
-        Supported on float64 operands only.
+        Supported on int32, int64, uint32, uint64, float32, float64 operands only.
+        """
+
+    class min(Stub):
+        """min(ary, idx, val)
```
Comment (pitrou) [inline numba/cuda/stubs.py]: The documentation here fails to mention the return value (same for `cuda.atomic.max()`).

File: numba/cuda/cudadrv/nvvm.py [hunk 6]
```diff
@@ -384,34 +384,65 @@ def get(self):
...
+    %cas = cmpxchg volatile {Ti}* %iptr, {Ti} %iold, {Ti} %ival monotonic
+    %dcas = bitcast {Ti} %cas to {T}
     br label %lt_check
 
 done:
     ; Return max
-    %ret = phi double [ %ptrval, %entry ], [ %dold, %lt_check ]
-    ret double %ret
-}
+    %ret = phi {T} [ %ptrval, %entry ], [ %dold, %lt_check ]
```
Comment (pitrou) [inline numba/cuda/cudadrv/nvvm.py]: I may be misunderstanding here, but `%dold` is the old value that was swapped out, that is, `%dold` is smaller than the value that was swapped in. How can it be the max, as commented above?

Comment (sklam) [inline numba/cuda/cudadrv/nvvm.py]: I should to add more description.  It is storing the max into the pointer and returning the old value.  All the atomic functions are like that.

**Comment by seibert** [comment]
Looks good now.  Merging.

**Pull request merged**